pkg/cloudprovider/provider/vsphere/helper.go

```diff
@@ -455,3 +456,47 @@ func resolveResourcePoolRef(ctx context.Context, config *Config, session *Sessio
 	}
 	return nil, nil
 }
+
+func createAndAttachTags(ctx context.Context, config *Config, vm *object.VirtualMachine) error {
+	restAPISession, err := NewRESTSession(ctx, config)
+	if err != nil {
+		return fmt.Errorf("failed to create REST API session: %v", err)
+	}
+	defer restAPISession.Logout(ctx)
+	tagManager := tags.NewManager(restAPISession.Client)
+	klog.V(3).Info("Creating tags")
+	for _, tag := range config.Tags {
+		tagID, err := tagManager.CreateTag(ctx, &tag)
+		if err != nil {
+			return fmt.Errorf("failed to create tag: %v %v", tag, err)
+		}
+
+		if err := tagManager.AttachTag(ctx, tagID, vm.Reference()); err != nil {
```

Comment:
Would it make sense to delete the tag if the attaching failed? I am looking at the way you delete the tags and what you do is you fetch all vm-referenced tags and then clean them, so if you failed to attach a tag here, then we will have orphan tags. 

I would suggest either those tags deleted here, or at the deletion stage, fetch the tags based on the machine not the vm and then clean them.